// env-runner/dist/_chunks/base-runner.mjs
import { rm } from "node:fs/promises";
import { proxyFetch, proxyUpgrade } from "httpxy";
var BaseEnvRunner = class {
	closed = false;
	_name;
	_workerEntry;
	_data;
	_hooks;
	_address;
	_messageListeners;
	constructor(opts) {
		this._name = opts.name;
		this._workerEntry = opts.workerEntry;
		this._data = opts.data;
		this._hooks = opts.hooks || {};
		this._messageListeners = /* @__PURE__ */ new Set();
	}
	get ready() {
		return Boolean(!this.closed && this._address && this._hasRuntime());
	}
	async fetch(input, init) {
		for (let i = 0; i < 5 && !this._address && !this.closed; i++) await new Promise((r) => setTimeout(r, 100 * Math.pow(2, i)));
		if (!this._address) return new Response(`${this._runtimeType()} env runner is unavailable`, { status: 503 });
		return proxyFetch(this._address, input, init);
	}
	async upgrade(context) {
		if (!this.ready || !this._address) return;
		await proxyUpgrade(this._address, context.node.req, context.node.socket, context.node.head);
	}
	onMessage(listener) {
		this._messageListeners.add(listener);
	}
	offMessage(listener) {
		this._messageListeners.delete(listener);
	}
	waitForReady(timeout = 5e3) {
		if (this.ready) return Promise.resolve();
		return new Promise((resolve, reject) => {
			const timer = setTimeout(() => {
				this._messageListeners.delete(listener);
				reject(/* @__PURE__ */ new Error("Runner did not become ready in time"));
			}, timeout);
			const listener = () => {
				if (this.ready) {
					clearTimeout(timer);
					this._messageListeners.delete(listener);
					resolve();
				}
			};
			this._messageListeners.add(listener);
		});
	}
	rpc(name, data, opts) {
		const id = Math.random().toString(36).slice(2);
		const timeout = opts?.timeout ?? 3e3;
		return new Promise((resolve, reject) => {
			const timer = setTimeout(() => {
				cleanup();
				reject(/* @__PURE__ */ new Error(`RPC "${name}" timed out`));
			}, timeout);
			const listener = (msg) => {
				if (msg?.__rpc_id === id) {
					cleanup();
					if (msg.error) reject(typeof msg.error === "string" ? new Error(msg.error) : msg.error);
					else resolve(msg.data);
				}
			};
			const cleanup = () => {
				clearTimeout(timer);
				this.offMessage(listener);
			};
			this.onMessage(listener);
			this.sendMessage({
				__rpc: name,
				__rpc_id: id,
				data
			});
		});
	}
	async reloadModule(timeout = 5e3) {
		return new Promise((resolve, reject) => {
			const timer = setTimeout(() => {
				cleanup();
				reject(/* @__PURE__ */ new Error("Module reload timed out"));
			}, timeout);
			const listener = (msg) => {
				if (msg?.event === "module-reloaded") {
					cleanup();
					if (msg.error) reject(typeof msg.error === "string" ? new Error(msg.error) : msg.error);
					else resolve();
				}
			};
			const cleanup = () => {
				clearTimeout(timer);
				this.offMessage(listener);
			};
			this.onMessage(listener);
			this.sendMessage({ event: "reload-module" });
		});
	}
	async close(cause) {
		if (this.closed) return;
		this.closed = true;
		this._hooks.onClose?.(this, cause);
		this._hooks = {};
		const onError = (error) => console.error(error);
		await this._closeRuntime().catch(onError);
		await this._closeSocket().catch(onError);
	}
	[Symbol.for("nodejs.util.inspect.custom")]() {
		const status = this.closed ? "closed" : this.ready ? "ready" : "pending";
		return `${this.constructor.name}#${this._name}(${status})`;
	}
	_handleMessage(message) {
		if (message?.address) {
			this._address = message.address;
			this._hooks.onReady?.(this, this._address);
		}
		for (const listener of this._messageListeners) listener(message);
	}
	async _closeSocket() {
		const socketPath = this._address?.socketPath;
		if (socketPath && socketPath[0] !== "\0" && !socketPath.startsWith(String.raw`\\.\\pipe`)) await rm(socketPath).catch(() => {});
		this._address = void 0;
	}
};
export { BaseEnvRunner as t };
