// env-runner/dist/_chunks/runner2.mjs
import { t as BaseEnvRunner } from "./base-runner.mjs";
import { n as parse, t as init } from "./libs/cjs-module-lexer.mjs";
import { createRequire } from "node:module";
import { proxyUpgrade } from "httpxy";
import { readFileSync } from "node:fs";
import { fileURLToPath } from "node:url";
import { basename, dirname, resolve } from "node:path";
import { resolveModulePath } from "exsolve";
const IPC_PATH = "/__env_runner_ipc";
const IPC_BINDING = "__ENV_RUNNER_IPC";
function generateWrapper(entryPath, opts) {
	const staticReExport = opts?.dynamicOnly ? "" : `export * from ${JSON.stringify(entryPath)};`;
	const explicitExports = opts?.dynamicOnly && opts.exports?.length ? opts.exports.map((name) => `export { ${name} } from ${JSON.stringify(entryPath)};`).join("\n") : "";
	const fetchBody = opts?.captureErrors ?? true ? `try {
      return await entryFetch(request, env, ctx);
    } catch (e) {
      const error = e instanceof Error ? e : new Error(String(e));
      const body = JSON.stringify({
        error: error.message,
        stack: error.stack,
        name: error.constructor?.name || "Error",
      });
      return new Response(body, {
        status: 500,
        headers: { "Content-Type": "application/json", "X-Env-Runner-Error": "1" },
      });
    }` : `return entryFetch(request, env, ctx);`;
	return `import __process from "node:process";
if (!globalThis.process) { globalThis.process = __process; }
${staticReExport}
${explicitExports}

const __IPC_PATH = "${IPC_PATH}";
const __IPC_BINDING = "${IPC_BINDING}";
const __entryPath = ${JSON.stringify(entryPath)};
let __userEntry;
let __ipcInitialized = false;
let __serverWs;
let __currentEnv;

async function __loadEntry(env, path) {
  globalThis.__ENV_RUNNER_UNSAFE_EVAL__ = env.__ENV_RUNNER_UNSAFE_EVAL__;
  const importFn = env.__ENV_RUNNER_UNSAFE_EVAL__.newAsyncFunction(
    "return await import(path)",
    "loadEntry",
    "path"
  );
  const mod = await importFn(path);
  return mod.default || mod;
}

function __sendMessage(message) {
  const payload = JSON.stringify(message);
  const env = __currentEnv;
  if (env && env[__IPC_BINDING]) {
    env[__IPC_BINDING].fetch("http://localhost/__ipc", {
      method: "POST",
      body: payload,
    }).catch(() => {});
    return;
  }
  if (__serverWs) {
    __serverWs.send(payload);
  }
}

async function __handleWsMessage(env, data) {
  let msg;
  try { msg = JSON.parse(data); } catch { return; }

  if (msg.type === "message") {
    if (__userEntry?.ipc?.onMessage) {
      __userEntry.ipc.onMessage(msg.data);
    }
    return;
  }

  if (msg.type === "reload" && env.__ENV_RUNNER_UNSAFE_EVAL__) {
    const version = msg.version || 0;
    try {
      const newEntry = await __loadEntry(env, __entryPath + "?t=" + version);
      if (__userEntry?.ipc?.onClose) {
        await __userEntry.ipc.onClose();
      }
      __userEntry = newEntry;
      __crosswsAdapter = undefined;
      __ipcInitialized = false;
      if (__userEntry.ipc?.onOpen) {
        __ipcInitialized = true;
        await __userEntry.ipc.onOpen({ sendMessage: __sendMessage });
      }
      __sendMessage({ event: "module-reloaded" });
    } catch (e) {
      __sendMessage({ event: "module-reloaded", error: String(e) });
    }
    return;
  }

  if (msg.type === "shutdown") {
    if (__userEntry?.ipc?.onClose) {
      await __userEntry.ipc.onClose();
    }
    return;
  }
}

let __crosswsAdapter;

async function __initCrossws(env, hooks) {
  if (__crosswsAdapter) return __crosswsAdapter;
  const importFn = env.__ENV_RUNNER_UNSAFE_EVAL__.newAsyncFunction(
    "return await import('crossws/adapters/cloudflare')",
    "loadCrossws"
  );
  const { default: cloudflareAdapter } = await importFn();
  __crosswsAdapter = cloudflareAdapter({ hooks });
  return __crosswsAdapter;
}

export default {
  async fetch(request, env, ctx) {
    const url = new URL(request.url);

    // WebSocket IPC handshake
    if (url.pathname === __IPC_PATH && request.headers.get("upgrade") === "websocket") {
      try {
        if (!__userEntry) {
          __userEntry = await __loadEntry(env, __entryPath);
        }
      } catch (e) {
        return new Response("Failed to load entry: " + String(e), { status: 500 });
      }

      const pair = new WebSocketPair();
      const client = pair[0];
      const server = pair[1];
      server.accept();
      __serverWs = server;

      server.addEventListener("message", (event) => {
        __handleWsMessage(env, event.data);
      });

      // Initialize IPC hooks
      if (!__ipcInitialized && __userEntry.ipc) {
        __ipcInitialized = true;
        if (__userEntry.ipc.onOpen) {
          await __userEntry.ipc.onOpen({ sendMessage: __sendMessage });
        }
      }

      return new Response(null, { status: 101, webSocket: client });
    }

    if (!__userEntry) {
      return new Response("Worker not initialized", { status: 503 });
    }

    // Handle WebSocket upgrade via crossws cloudflare adapter
    if (__userEntry.websocket && request.headers.get("upgrade") === "websocket") {
      const adapter = await __initCrossws(env, __userEntry.websocket);
      return adapter.handleUpgrade(request, env, ctx);
    }

    const entryFetch = __userEntry.fetch;
    if (!entryFetch) {
      return new Response("No fetch handler exported", { status: 500 });
    }
    __currentEnv = env;
    try {
      ${fetchBody}
    } finally {
      __currentEnv = undefined;
    }
  }
};
`;
}
const _miniflareCache = /* @__PURE__ */ new Map();
var MiniflareEnvRunner = class extends BaseEnvRunner {
	#miniflare;
	#miniflareOptions;
	#transformRequest;
	#reloadCounter = 0;
	#ws;
	#persistent;
	#cacheKey;
	#exports;
	#captureErrors;
	#exportConditions;
	constructor(opts) {
		super({
			...opts,
			workerEntry: ""
		});
		this.#miniflareOptions = opts.miniflareOptions || {};
		this.#transformRequest = opts.transformRequest;
		this.#persistent = opts.persistent ?? false;
		this.#exports = opts.exports ?? {};
		this.#captureErrors = opts.captureErrors ?? true;
		this.#exportConditions = opts.exportConditions ?? ["workerd", "worker"];
		this.#init();
	}
	static async disposeAll() {
		const entries = [..._miniflareCache.values()];
		_miniflareCache.clear();
		for (const entry of entries) await entry.mf.dispose().catch(() => {});
	}
	async dispose() {
		if (this.#miniflare) {
			if (this.#ws) {
				this.#ws.send(JSON.stringify({ type: "shutdown" }));
				this.#ws.close();
				this.#ws = void 0;
			}
			if (this.#cacheKey) _miniflareCache.delete(this.#cacheKey);
			await this.#miniflare.dispose();
			this.#miniflare = void 0;
		}
		if (!this.closed) await this.close();
	}
	async fetch(input, init) {
		if (!this.#miniflare || this.closed) return new Response("miniflare env runner is unavailable", { status: 503 });
		const url = typeof input === "string" ? input : input instanceof URL ? input.href : input.url;
		const res = await this.#miniflare.dispatchFetch(url, init);
		if (res instanceof Response) return res;
		return new Response(res.body, {
			status: res.status,
			statusText: res.statusText,
			headers: res.headers
		});
	}
	sendMessage(message) {
		if (!this.#ws) throw new Error("Miniflare env runner should be initialized before sending messages.");
		if (message?.type === "ping") {
			queueMicrotask(() => this._handleMessage({
				type: "pong",
				data: message.data
			}));
			return;
		}
		this.#ws.send(JSON.stringify({
			type: "message",
			data: message
		}));
	}
	async reloadModule(timeout = 5e3) {
		if (!this.#ws) throw new Error("Miniflare env runner should be initialized before reloading.");
		if (!this._data?.entry) return;
		this.#reloadCounter++;
		const version = this.#reloadCounter;
		return new Promise((resolve, reject) => {
			const timer = setTimeout(() => {
				cleanup();
				reject(/* @__PURE__ */ new Error("Module reload timed out"));
			}, timeout);
			const listener = (msg) => {
				if (msg?.event === "module-reloaded") {
					cleanup();
					if (msg.error) reject(typeof msg.error === "string" ? new Error(msg.error) : msg.error);
					else resolve();
				}
			};
			const cleanup = () => {
				clearTimeout(timer);
				this.offMessage(listener);
			};
			this.onMessage(listener);
			this.#ws.send(JSON.stringify({
				type: "reload",
				version
			}));
		});
	}
	_hasRuntime() {
		return Boolean(this.#miniflare);
	}
	_runtimeType() {
		return "miniflare";
	}
	async _closeRuntime() {
		if (!this.#miniflare) return;
		if (this.#ws) {
			this.#ws.send(JSON.stringify({ type: "shutdown" }));
			this.#ws.close();
			this.#ws = void 0;
		}
		if (this.#persistent && this.#cacheKey) {
			const cached = _miniflareCache.get(this.#cacheKey);
			if (cached) {
				cached.refCount--;
				if (cached.refCount <= 0) {
					_miniflareCache.delete(this.#cacheKey);
					await this.#miniflare.dispose();
				}
			}
		} else await this.#miniflare.dispose();
		this.#miniflare = void 0;
	}
	#init() {
		this.#initAsync().catch((error) => {
			console.error("Miniflare runner init error:", error);
			this.close(error);
		});
	}
	async upgrade(context) {
		if (!this.#miniflare || this.closed) {
			context.node.socket.destroy();
			return;
		}
		const mfUrl = await this.#miniflare.unsafeGetDirectURL();
		const address = new URL(mfUrl);
		await proxyUpgrade({
			host: address.hostname,
			port: Number(address.port)
		}, context.node.req, context.node.socket, context.node.head);
	}
	async #initAsync() {
		const { Miniflare } = await import("miniflare");
		const entryPath = this._data?.entry;
		const userFlags = this.#miniflareOptions.compatibilityFlags || [];
		const userDirectSockets = this.#miniflareOptions.unsafeDirectSockets || [];
		const options = {
			compatibilityDate: (/* @__PURE__ */ new Date()).toISOString().split("T")[0],
			modules: true,
			...this.#miniflareOptions,
			compatibilityFlags: [...new Set(["nodejs_compat", ...userFlags])],
			unsafeDirectSockets: [{
				host: "127.0.0.1",
				port: 0
			}, ...userDirectSockets]
		};
		if (entryPath && !options.script && !options.scriptPath) {
			const resolvedEntry = resolve(entryPath);
			const entryDir = dirname(resolvedEntry);
			const detectedExports = this.#exports === false || this.#exports === void 0 ? [] : detectExportedClasses(resolvedEntry, typeof this.#exports === "object" ? this.#exports : {});
			if (detectedExports.length > 0 && !options.durableObjects) {
				const autoDOs = { ...this.#miniflareOptions.durableObjects || {} };
				for (const name of detectedExports) {
					const bindingName = toScreamingSnakeCase(name);
					if (!autoDOs[bindingName]) autoDOs[bindingName] = name;
				}
				options.durableObjects = autoDOs;
			}
			options.script = generateWrapper(resolvedEntry, {
				dynamicOnly: true,
				captureErrors: this.#captureErrors,
				exports: detectedExports
			});
			options.scriptPath = entryDir + "/__env_runner_wrapper.mjs";
			if (!options.modulesRoot) options.modulesRoot = "/";
			options.unsafeEvalBinding = "__ENV_RUNNER_UNSAFE_EVAL__";
			options.serviceBindings = {
				...options.serviceBindings || {},
				[IPC_BINDING]: async (request) => {
					try {
						const message = await request.json();
						this._handleMessage(message);
					} catch {}
					return new Response(null, { status: 204 });
				}
			};
			if (this.#transformRequest && !options.modulesRules) options.modulesRules = [{
				type: "ESModule",
				include: [
					"**/*.ts",
					"**/*.tsx",
					"**/*.jsx",
					"**/*.mts"
				]
			}];
			if (!options.unsafeModuleFallbackService) {
				const _require = createRequire(resolvedEntry);
				const _transformRequest = this.#transformRequest;
				const _exportConditions = this.#exportConditions;
				options.unsafeUseModuleFallbackService = true;
				const modulePathMap = /* @__PURE__ */ new Map();
				const _cjsLexerReady = ensureCjsLexer();
				options.unsafeModuleFallbackService = async (request) => {
					await _cjsLexerReady;
					const url = new URL(request.url);
					const specifier = url.searchParams.get("specifier");
					const rawSpecifier = url.searchParams.get("rawSpecifier");
					const referrer = url.searchParams.get("referrer") || "";
					if (!specifier) return new Response(null, { status: 404 });
					const cleanSpecifier = specifier.split("?")[0] || specifier;
					const cleanRaw = rawSpecifier?.split("?")[0];
					let resolvedPath;
					const fileUrlRaw = cleanRaw || cleanSpecifier;
					if (fileUrlRaw.startsWith("file://")) try {
						resolvedPath = fileURLToPath(fileUrlRaw);
					} catch {
						return new Response(null, { status: 404 });
					}
					else if (cleanRaw && !cleanRaw.startsWith(".") && !cleanRaw.startsWith("/")) {
						const referrerKey = referrer.startsWith("/") ? referrer.slice(1) : referrer;
						const referrerReal = modulePathMap.get(referrerKey);
						const contextRequire = referrerReal ? createRequire(referrerReal) : _require;
						if (cleanRaw.startsWith("cloudflare:")) return new Response(null, { status: 404 });
						if (cleanRaw.startsWith("node:")) {
							const nodeName = cleanRaw.slice(5);
							try {
								resolvedPath = contextRequire.resolve(`unenv/node/${nodeName}`);
							} catch {
								return new Response(null, { status: 404 });
							}
						} else try {
							resolvedPath = resolveModulePath(cleanRaw, {
								from: referrerReal || resolvedEntry,
								conditions: _exportConditions,
								try: true
							}) || contextRequire.resolve(cleanRaw);
						} catch {
							const name = cleanSpecifier.startsWith("/") ? cleanSpecifier.slice(1) : cleanSpecifier;
							return Response.json({
								name,
								esModule: "export default undefined;"
							});
						}
					} else {
						const referrerKey = referrer.startsWith("/") ? referrer.slice(1) : referrer;
						const referrerDir = dirname(modulePathMap.get(referrerKey) || (referrer.startsWith("/") ? referrer : "/" + referrer));
						const raw = cleanRaw || cleanSpecifier;
						if (raw.startsWith(".")) resolvedPath = resolve(referrerDir, raw);
						else if (cleanSpecifier.startsWith("/")) resolvedPath = cleanSpecifier;
						else try {
							resolvedPath = _require.resolve(raw);
						} catch {
							return new Response(null, { status: 404 });
						}
					}
					const rawQuery = specifier.includes("?") ? specifier.slice(specifier.indexOf("?")) : "";
					const name = (cleanSpecifier.startsWith("/") ? cleanSpecifier.slice(1) : cleanSpecifier) + rawQuery;
					if (_transformRequest) try {
						const result = await _transformRequest(resolvedPath);
						if (result?.code) {
							modulePathMap.set(name, resolvedPath);
							return Response.json({
								name,
								esModule: result.code
							});
						}
					} catch {}
					try {
						const contents = readFileSync(resolvedPath, "utf8");
						modulePathMap.set(name, resolvedPath);
						if (resolvedPath.endsWith(".mjs") || !resolvedPath.endsWith(".cjs") && /\b(import\s|import\(|export\s|export\{|import\.meta\b)/.test(contents)) return Response.json({
							name,
							esModule: contents
						});
						const cjsSuffix = "?__cjs";
						if (specifier.endsWith(cjsSuffix)) return Response.json({
							name,
							commonJsModule: contents
						});
						const esModule = createCjsEsmShim("./" + basename(resolvedPath) + cjsSuffix, contents);
						return Response.json({
							name,
							esModule
						});
					} catch {
						return new Response(null, { status: 404 });
					}
				};
			}
		}
		if (this.#persistent && entryPath) {
			this.#cacheKey = computeCacheKey(entryPath, {
				...this.#miniflareOptions,
				_exportConditions: this.#exportConditions
			});
			const cached = _miniflareCache.get(this.#cacheKey);
			if (cached) {
				this.#miniflare = cached.mf;
				cached.refCount++;
			}
		}
		if (!this.#miniflare) {
			this.#miniflare = new Miniflare(options);
			await this.#miniflare.ready;
			if (this.#persistent && this.#cacheKey) _miniflareCache.set(this.#cacheKey, {
				mf: this.#miniflare,
				refCount: 1
			});
		}
		const initRes = await this.#miniflare.dispatchFetch("http://localhost/__env_runner_ipc", { headers: { upgrade: "websocket" } });
		const ws = initRes.webSocket;
		if (!ws) {
			const body = await initRes.text().catch(() => "");
			throw new Error(`Failed to establish WebSocket IPC channel (${initRes.status}: ${body})`);
		}
		ws.accept();
		this.#ws = ws;
		ws.addEventListener("message", (event) => {
			try {
				const parsed = JSON.parse(event.data);
				this._handleMessage(parsed);
			} catch {}
		});
		this._handleMessage({ address: {
			host: "127.0.0.1",
			port: 0
		} });
	}
};
function detectExportedClasses(entryPath, explicit) {
	const names = new Set(Object.keys(explicit));
	try {
		const source = readFileSync(entryPath, "utf8");
		const re = /\bexport\s+class\s+(\w+)/g;
		let match;
		while (match = re.exec(source)) if (match[1]) names.add(match[1]);
	} catch {}
	return [...names];
}
function toScreamingSnakeCase(name) {
	return name.replace(/([a-z0-9])([A-Z])/g, "$1_$2").toUpperCase();
}
function computeCacheKey(entryPath, opts) {
	const serializableOpts = {};
	for (const [k, v] of Object.entries(opts)) if (typeof v !== "function") serializableOpts[k] = v;
	return `${resolve(entryPath)}::${JSON.stringify(serializableOpts)}`;
}
let _cjsLexerReady;
function ensureCjsLexer() {
	if (!_cjsLexerReady) _cjsLexerReady = init();
	return _cjsLexerReady;
}
function createCjsEsmShim(cjsSpecifier, contents) {
	let namedExports = [];
	try {
		const { exports } = parse(contents);
		namedExports = exports.filter((e) => e !== "default" && e !== "__esModule");
	} catch {}
	let shim = `import __cjs_mod__ from ${JSON.stringify(cjsSpecifier)};\nexport default __cjs_mod__;\n`;
	for (const name of namedExports) shim += `export var ${name} = __cjs_mod__["${name}"];\n`;
	return shim;
}
export { MiniflareEnvRunner as t };
