// env-runner/dist/_chunks/worker-utils.mjs
import { readFileSync } from "node:fs";
import { pathToFileURL } from "node:url";
import { isAbsolute } from "node:path";
async function resolveEntry(entryPath) {
	const mod = await import(_toImportPath(entryPath));
	const entry = mod.default || mod;
	if (typeof entry.fetch !== "function") throw new Error(`[env-runner] Entry module "${entryPath}" must export a \`fetch\` handler (export default { fetch(req) { ... } }).`);
	return entry;
}
function parseServerAddress(server) {
	const url = new URL(server.url);
	return {
		host: url.hostname,
		port: Number(url.port)
	};
}
async function reloadEntryModule(entryPath, currentEntry, sendMessage) {
	await currentEntry.ipc?.onClose?.();
	const newEntry = await _importFresh(entryPath);
	await newEntry.ipc?.onOpen?.({ sendMessage });
	return newEntry;
}
function _toImportPath(entryPath) {
	const qIndex = entryPath.indexOf("?");
	const filePath = qIndex === -1 ? entryPath : entryPath.slice(0, qIndex);
	const query = qIndex === -1 ? "" : entryPath.slice(qIndex);
	if (isAbsolute(filePath)) return pathToFileURL(filePath).href + query;
	return entryPath;
}
async function _importFresh(entryPath) {
	const code = readFileSync(entryPath, "utf8");
	const mod = await import("data:text/javascript;base64," + Buffer.from(code).toString("base64"));
	const entry = mod.default || mod;
	if (typeof entry.fetch !== "function") throw new Error(`[env-runner] Entry module "${entryPath}" must export a \`fetch\` handler (export default { fetch(req) { ... } }).`);
	return entry;
}
export { reloadEntryModule as n, resolveEntry as r, parseServerAddress as t };
