// @tanstack/router-utils/dist/cjs/index.cjs
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const require_ast = require("./ast.cjs");
const require_logger = require("./logger.cjs");
const require_copy_files_plugin = require("./copy-files-plugin.cjs");
const require_path_ids = require("./path-ids.cjs");
const require_compiler_helpers = require("./compiler-helpers.cjs");
let babel_dead_code_elimination = require("babel-dead-code-elimination");
exports.buildDeclarationMap = require_compiler_helpers.buildDeclarationMap;
exports.buildDependencyGraph = require_compiler_helpers.buildDependencyGraph;
exports.collectIdentifiersFromNode = require_compiler_helpers.collectIdentifiersFromNode;
exports.collectIdentifiersFromPattern = require_compiler_helpers.collectIdentifiersFromPattern;
exports.collectLocalBindingsFromStatement = require_compiler_helpers.collectLocalBindingsFromStatement;
exports.collectModuleLevelRefsFromNode = require_compiler_helpers.collectModuleLevelRefsFromNode;
exports.copyFilesPlugin = require_copy_files_plugin.copyFilesPlugin;
exports.createIdentifier = require_path_ids.createIdentifier;
exports.deadCodeElimination = require_ast.deadCodeElimination;
exports.decodeIdentifier = require_path_ids.decodeIdentifier;
exports.expandDestructuredDeclarations = require_compiler_helpers.expandDestructuredDeclarations;
exports.expandSharedDestructuredDeclarators = require_compiler_helpers.expandSharedDestructuredDeclarators;
exports.expandTransitively = require_compiler_helpers.expandTransitively;
exports.extractModuleInfoFromAst = require_compiler_helpers.extractModuleInfoFromAst;
exports.findReferencedIdentifiers = babel_dead_code_elimination.findReferencedIdentifiers;
exports.generateFromAst = require_ast.generateFromAst;
exports.getVariableDeclaratorForExpressionPath = require_compiler_helpers.getVariableDeclaratorForExpressionPath;
exports.logDiff = require_logger.logDiff;
exports.parseAst = require_ast.parseAst;
exports.removeBindingsTransitivelyDependingOn = require_compiler_helpers.removeBindingsTransitivelyDependingOn;
exports.removeModuleLevelBindings = require_compiler_helpers.removeModuleLevelBindings;
exports.retainModuleLevelDeclarations = require_compiler_helpers.retainModuleLevelDeclarations;
exports.stripTypeExports = require_ast.stripTypeExports;
exports.stripUnreferencedTopLevelExpressionStatements = require_compiler_helpers.stripUnreferencedTopLevelExpressionStatements;
exports.unwrapExportedDeclarations = require_compiler_helpers.unwrapExportedDeclarations;
exports.unwrapExpression = require_compiler_helpers.unwrapExpression;
