// env-runner/dist/runners/deno-process/worker.mjs
import { n as reloadEntryModule, r as resolveEntry, t as parseServerAddress } from "../../_chunks/worker-utils.mjs";
import { serve } from "srvx";
import { plugin } from "crossws/server";
const data = JSON.parse(process.env.ENV_RUNNER_DATA || "{}");
let entry = await resolveEntry(data.entry);
const _stdout = globalThis.Deno?.stdout ? { write: (s) => globalThis.Deno.stdout.writeSync(new TextEncoder().encode(s)) } : process.stdout;
const sendMessage = (message) => _stdout.write(JSON.stringify(message) + "\n");
const _stdin = globalThis.Deno?.stdin?.readable || process.stdin;
const server = serve({
	port: 0,
	hostname: "127.0.0.1",
	silent: true,
	fetch: (request) => entry.fetch(request),
	middleware: entry.middleware,
	plugins: [...entry.plugins || [], ...entry.websocket ? [plugin(entry.websocket)] : []],
	gracefulShutdown: false
});
await server.ready();
if (entry.upgrade) server.node?.server?.on("upgrade", (req, socket, head) => {
	entry.upgrade({ node: {
		req,
		socket,
		head
	} });
});
if (entry.ipc) await entry.ipc.onOpen?.({ sendMessage });
sendMessage({ address: parseServerAddress(server) });
async function readMessages() {
	const decoder = new TextDecoder();
	let buffer = "";
	for await (const chunk of _stdin) {
		buffer += decoder.decode(chunk, { stream: true });
		let newlineIdx;
		while ((newlineIdx = buffer.indexOf("\n")) !== -1) {
			const line = buffer.slice(0, newlineIdx);
			buffer = buffer.slice(newlineIdx + 1);
			if (!line) continue;
			let message;
			try {
				message = JSON.parse(line);
			} catch {
				continue;
			}
			await handleMessage(message);
		}
	}
}
async function handleMessage(message) {
	if (message?.event === "shutdown") {
		Promise.resolve(entry.ipc?.onClose?.()).then(() => server.close()).then(() => {
			sendMessage({ event: "exit" });
		});
		return;
	}
	if (message?.event === "reload-module") {
		try {
			entry = await reloadEntryModule(data.entry, entry, sendMessage);
			sendMessage({ event: "module-reloaded" });
		} catch (error) {
			sendMessage({
				event: "module-reloaded",
				error: error?.message || String(error)
			});
		}
		return;
	}
	if (message?.type === "ping") {
		sendMessage({
			type: "pong",
			data: message.data
		});
		return;
	}
	entry.ipc?.onMessage?.(message);
}
readMessages().catch(() => {});
export {};
